gobblin-data-management/src/main/java/gobblin/data/management/retention/dataset/DatasetBase.java

```diff
@@ -0,0 +1,215 @@
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.data.management.retention.dataset;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Lists;
+
+import azkaban.utils.Props;
+
+import gobblin.data.management.retention.version.DatasetVersion;
+import gobblin.data.management.retention.policy.RetentionPolicy;
+import gobblin.data.management.retention.version.finder.VersionFinder;
+import gobblin.data.management.trash.Trash;
+
+
+/**
+ * Implementation of a {@link gobblin.data.management.retention.dataset.Dataset} that uses a
+ * {@link gobblin.data.management.retention.version.finder.VersionFinder} to find dataset versions, a
+ * {@link gobblin.data.management.retention.policy.RetentionPolicy} to figure out deletable versions, and then deletes
+ * those files and newly empty parent directories.
+ *
+ * <p>
+ *   Concrete subclasses should implement {@link #getVersionFinder} and {@link #getRetentionPolicy}.
+ * </p>
+ *
+ * <p>
+ * Datasets are directories in the filesystem containing data files organized in version-like directory structures.
+ * Example datasets:
+ * </p>
+ *
+ * <p>
+ *   For snapshot based datasets, with the directory structure:
+ *   <pre>
+ *   /path/to/table/
+ *      snapshot1/
+ *          dataFiles...
+ *      snapshot2/
+ *          dataFiles...
+ *   </pre>
+ *   each of snapshot1 and snapshot2 are dataset versions.
+ * </p>
+ *
+ * <p>
+ *   For tracking datasets, with the directory structure:
+ *   <pre>
+ *   /path/to/tracking/data/
+ *      2015/
+ *          06/
+ *              01/
+ *                  dataFiles...
+ *              02/
+ *                  dataFiles...
+ *   </pre>
+ *   each of 2015/06/01 and 2015/06/02 are dataset versions.
+ * </p>
+ *
+ * <p>
+ *  {@link DatasetBase} uses a {@link gobblin.data.management.retention.version.finder.DatasetVersionFinder} to find all
+ *  subdirectories that are versions of this dataset. After that, for each dataset, it uses a
+ *  {@link gobblin.data.management.retention.policy.RetentionPolicy} to decide which versions of the dataset should be
+ *  deleted. For each version deleted, if {@link #deleteEmptyDirectories} it will also look at all parent directories
+ *  and delete directories that are now empty, up to but not including the dataset root.
+ * </p>
+ *
+ * @param <T> type of {@link gobblin.data.management.retention.version.DatasetVersion} supported by this
+ *           {@link gobblin.data.management.retention.dataset.Dataset}.
+ */
+public abstract class DatasetBase<T extends DatasetVersion> implements Dataset {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DatasetBase.class);
```

Comment:
Please use a non-static logger. You may still use the static version as a default but allow for callers to pass their category name. This will allow use in parallel multiple cleaners for different datasets and still now which cleaner logged each message.